File: 设计数据结构/LRU缓存机制.py
# 146.LRU缓存机制（中等）
# 双链表+哈希表
class DLinkedNode:
    def __init__(self, key=0, value=0):
        self.key = key
        self.val = value
        self.next = None
        self.prev = None

class LRUCache:
    def __init__(self, capacity):
        self.cache = dict()
        self.head = DLinkedNode()
        self.tail = DLinkedNode()
        self.head.next = self.tail
        self.tail.prev = self.head
        self.capacity = capacity
        self.size = 0

    def get(self, key):
        if key not in self.cache:
            return -1
        query_node = self.cache[key]  # 哈希表获取要查询的链表节点

        self.moveToTail(query_node)
        return query_node.val
    
    def put(self, key, value):  
        if key not in self.cache:
            add_node = DLinkedNode(key=key, value=value)

            self.moveToTail(add_node)  # 新节点插入到尾部

            self.cache[key] = add_node  # 添加至缓存中
            self.size += 1
            if self.size > self.capacity:  # 超出缓存，则删除头部节点
                remove_node = self.head.next  # 待删除节点
                self.head.next.next.prev = self.head
                self.head.next = self.head.next.next

                self.size -= 1
                self.cache.pop(remove_node.key)
        else:
            modify_node = self.cache[key]
            modify_node.val = value  # 修改指定节点
            modify_node.prev.next = modify_node.next  
            modify_node.next.prev = modify_node.prev

            self.moveToTail(modify_node)

    def moveToTail(self, node):
        node.next = self.tail  # 移至尾部
        node.prev = self.tail.prev
        self.tail.prev.next = node
        self.tail.prev = node
    
    # def removeNode(self, node):

LRU = LRUCache(2)
LRU.put(2,1)
LRU.put(2,2)
print(LRU.get(2))
LRU.put(1,3)
LRU.put(3,4)
print(LRU.get(2))

File: 设计数据结构/LRU缓存机制.java
package com.company;

import java.util.HashMap;
import java.util.Map;

public class LRUCache {
    class DLinkNode{
        int key;
        int value;
        DLinkNode prev;
        DLinkNode next;
        DLinkNode(){}
        DLinkNode(int key, int value){
            this.key = key;
            this.value = value;
        }
    }

    private Map<Integer, DLinkNode> cache = new HashMap<Integer, DLinkNode>();
    private int size;  // 当前容量使用情况
    private int capacity;  // 拟定容量大小
    private DLinkNode head, tail; // 伪头、尾节点

    /**
     *  LRUCache
     *  构造函数初始化
     */
    public LRUCache(int capacity){
        this.capacity = capacity;
        head = new DLinkNode();  // 伪头节点
        tail = new DLinkNode();  // 伪尾节点
        head.next = tail;
        tail.prev = head;
    }

    /**
     * 根据key返回缓存中对应的值
     * @param key
     * @return
     */
    public int get(int key){
        DLinkNode node = cache.get(key);
        if (node != null){
            // key存在，通过哈希表定位后移至头部
            moveToHead(node);
            return node.value;
        }else {
            return -1;
        }

    }

    public void put(int key, int value){
        DLinkNode node = cache.get(key);
        if (node != null){  // 当缓存中已存在key，改变值
            node.value = value;
            moveToHead(node);
        }else {
            // 不存在，则加入新节点
            DLinkNode newNode = new DLinkNode(key, value);
            cache.put(key, newNode);
            addNewNode(newNode);
            if ( this.size > this.capacity ){
                // 超出缓存大小，移除最后一个节点
                DLinkNode removeNode = removeTail();
                cache.remove(removeNode.key);
                this.size --;
            }
        }
    }

    private void addNewNode(DLinkNode node){
        addToHead(node);
        this.size ++;
    }

    private void moveToHead(DLinkNode node){
        // 先把该节点移除出来
        removeNode(node);
        // 再将该节点添加至头部
        addToHead(node);
    }

    private void removeNode(DLinkNode node){
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }

    private void addToHead(DLinkNode node){
        DLinkNode tempNode = head.next;
        head.next = node;
        node.prev = head;
        node.next = tempNode;
        tempNode.prev = node;

    }

    private DLinkNode removeTail(){
        DLinkNode node = tail.prev;
        removeNode(node);
        return node;
    }

}
